pandas/core/case_when.py

```diff
@@ -0,0 +1,209 @@
+        if not isinstance(default, ABCDataFrame):
+            default = convert_to_list_like(default)
+        if not hasattr(default, "shape"):
+            default = pd_array(default, copy=False)
+        if default.ndim > 1:
+            raise ValueError(
+                "The provided default argument should "
+                "either be a scalar or a 1-D array."
+            )
+        if default.size != conditions[0].size:
+            raise ValueError(
+                "The length of the default argument should "
+                "be the same as the length of any "
+                "of the boolean conditions."
+            )
+    replacements = []
+    # ideally we could skip these checks and let `Series.mask`
+    # handle it - however, this step is necessary to get
+    # a common dtype for multiple conditions and replacements.
+    for num, replacement in zip(counter, args[-1::-2]):
+        if is_scalar(replacement):
+            replacement = construct_1d_arraylike_from_scalar(
+                replacement, length=bool_length, dtype=None
+            )
+        if not isinstance(replacement, ABCDataFrame):
+            replacement = convert_to_list_like(replacement)
+        if not hasattr(replacement, "shape"):
+            replacement = pd_array(replacement, copy=False)
+        if replacement.ndim > 1:
+            raise ValueError(f"replacement{num} should be a 1-D array.")
+        if replacement.size != bool_length:
+            raise ValueError(
+                f"The size of replacement{num} array"
+                f"does not match the size of condition{num} array."
+            )
+        replacements.append(replacement)
+    common_dtype = [arr.dtype for arr in replacements]
+    if default is not None:
+        common_dtype.append(default.dtype)
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            arr.astype(common_dtype, copy=False)
+            if isinstance(arr, ABCSeries)
+            else pd_array(arr, dtype=common_dtype, copy=False)
+            for arr in replacements
+        ]
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtype, copy=False)
+        elif default is not None:
+            default = pd_array(default, dtype=common_dtype, copy=False)
+    else:
+        common_dtype = common_dtype[0]
+
+    if default is None:
+        default = construct_1d_arraylike_from_scalar(
+            na_value_for_dtype(common_dtype, compat=False),
+            length=bool_length,
+            dtype=common_dtype,
+        )
+        default = Series(default)
+    elif not isinstance(default, ABCSeries):
+        default = Series(default)
+
+    for position, condition, replacement in zip(counter, conditions, replacements):
+        try:
+            default = default.mask(
+                condition, other=replacement, axis=0, inplace=False, level=level
+            )
```

Comment:
If default is passed as non-pandas object but the conditions / replacements are Series with a non-RangeIndex, I think the output will be all NA values.

Reply:
A common case where I think this goes wrong is `case_when(df["a"] > 1, df["b"], default=5)` where `df` has a non-RangeIndex.

Reply:
Thanks @rhshadrach for the review. I had a look at the concat source code but could not identify the alignment for axis=1. Could you kindly point me in the direction? I took it for granted that mask already does this alignment so isn't this sufficient?

Reply:
an alternative would be to change the current logic - if it is a scalar, grab the type and pass to `find_common_dtype`, do not turn it to an array - let masking, via alignment handle that. list like can stay with the current logic

Reply:
i dont think we should support outer join, since the replacement values should not increase the size of the original Series. The familiar implementation in SQL does not expand the column, and I do not see a good reason for this to do so either.

Reply:
note to self - add test for non range Index